truffle/src/main/java/org/jruby/truffle/format/parser/PrintfTreeBuilder.java

```diff
@@ -0,0 +1,240 @@
+
+        for (org.jruby.truffle.format.parser.PrintfParser.FlagContext flag : ctx.flag()) {
+            if (flag.MINUS() != null) {
+                leftJustified = true;
+            } else if (flag.SPACE() != null) {
+                spacePadding = width;
+            } else if (flag.ZERO() != null) {
+                zeroPadding = width;
+            } else {
+                throw new UnsupportedOperationException();
+            }
+        }
+
+        if (spacePadding == DEFAULT && zeroPadding == DEFAULT) {
+            spacePadding = width;
+        }
+
+        final char type = ctx.TYPE().getSymbol().getText().charAt(0);
+
+        final PackNode valueNode;
+
+        if (ctx.ANGLE_KEY() == null) {
+            valueNode = ReadValueNodeGen.create(context, new SourceNode());
+        } else {
+            final ByteList keyBytes = tokenAsBytes(ctx.ANGLE_KEY().getSymbol(), 1);
+            final DynamicObject key = context.getSymbol(keyBytes);
+            valueNode = ReadHashValueNodeGen.create(context, key, new SourceNode());
+        }
+
+        final int precision;
+
+        if (ctx.precision != null) {
+            precision = Integer.parseInt(ctx.precision.getText());
+        } else {
+            precision = DEFAULT;
+        }
+
+        final PackNode node;
+
+        switch (type) {
+            case 's':
+                if (ctx.ANGLE_KEY() == null) {
+                    if (spacePadding == DEFAULT) {
+                        node = WriteBytesNodeGen.create(context, ReadStringNodeGen.create(
+                                context, true, "to_s", false, new ByteList(), new SourceNode()));
+                    } else {
+                        node = WritePaddedBytesNodeGen.create(context, spacePadding, leftJustified,
+                                ReadStringNodeGen.create(context, true, "to_s", false, new ByteList(), new SourceNode()));
+                    }
+                } else {
+                    if (spacePadding == DEFAULT) {
+                        node = WriteBytesNodeGen.create(context, ToStringNodeGen.create(
+                                context, true, "to_s", false, new ByteList(), valueNode));
+                    } else {
+                        node = WritePaddedBytesNodeGen.create(context, spacePadding, leftJustified,
+                                ToStringNodeGen.create(context, true, "to_s", false, new ByteList(), valueNode));
+                    }
+                }
+                break;
+            case 'd':
+            case 'i':
+            case 'o':
+            case 'u':
+            case 'x':
+            case 'X':
+                final PackNode spacePaddingNode;
+                if (spacePadding == PADDING_FROM_ARGUMENT) {
+                    spacePaddingNode = ReadIntegerNodeGen.create(context, new SourceNode());
+                } else {
+                    spacePaddingNode = new LiteralIntegerNode(context, spacePadding);
+                }
```

Comment:
I hoped this would go away with the parsing if we had sth like
('*' => ReadPaddingFromArg | \d+ => LiteralPadding | '' NoPadding) but it might be hard.
Padding from arg should be supported for all formats.